### ISSUE TITLE
cpu_show_meltdown() is wrong when rfi_flush is disabled

### ISSUE THREAD

**Issue by mpe** [linuxppc/issues#243]
If we're on a system that requires the RFI flush, but the RFI flush was disabled, eg via debugfs, then we should report that we're vulnerable to Meltdown.

Currently we only do that if `SEC_FTR_L1D_THREAD_PRIV` is **also** false.

ie, currently this happens:
```
# cat /sys/devices/system/cpu/vulnerabilities/meltdown
Mitigation: RFI Flush, L1D private per thread
# echo 0 > /sys/kernel/debug/powerpc/rfi_flush
# cat /sys/devices/system/cpu/vulnerabilities/meltdown
Mitigation: L1D private per thread
```

But at the end it should say:
```
Vulnerable, L1D private per thread
```

In practice this probably only happens on Power9 systems.

**Comment by walbon** [comment]
Do you mean it must show "Vulnerable" if disable the "RFI Flush" just?
If "RFI Flush" is able and "L1D private per thread" is disable, it must be "Vulnerable" too?

**Comment by walbon** [comment]
```--- a/arch/powerpc/kernel/security.c
+++ b/arch/powerpc/kernel/security.c
@@ -130,26 +130,16 @@ ssize_t cpu_show_meltdown(struct device *dev, struct device_attribute *attr, cha

        thread_priv = security_ftr_enabled(SEC_FTR_L1D_THREAD_PRIV);

-       if (rfi_flush || thread_priv) {
-               struct seq_buf s;
-               seq_buf_init(&s, buf, PAGE_SIZE - 1);
-
-               seq_buf_printf(&s, "Mitigation: ");
-
-               if (rfi_flush)
-                       seq_buf_printf(&s, "RFI Flush");
-
-               if (rfi_flush && thread_priv)
-                       seq_buf_printf(&s, ", ");
-
+       if (rfi_flush) {
                if (thread_priv)
-                       seq_buf_printf(&s, "L1D private per thread");
-
-               seq_buf_printf(&s, "\n");
-
-               return s.len;
+                       return sprintf(buf, "Mitigation: RFI Flush, L1D private per thread\n");
+               else
+                       return sprintf(buf, "Vulnerable: L1D private per thread\n");
        }

+       if (thread_priv)
+               return sprintf(buf, "Vulnerable: L1D private per thread\n");
+
        if (!security_ftr_enabled(SEC_FTR_L1D_FLUSH_HV) &&
            !security_ftr_enabled(SEC_FTR_L1D_FLUSH_PR))
                return sprintf(buf, "Not affected\n");
```

**Comment by walbon** [comment]
I created a table to show my doubt in the case of the P8, which it has not the feature of  `L1D per thread` and whenever it should show `Vulnerable`.

CPU | L1D per Thread | RFI Flush | Vulnerable
----| ----| --- | ---
P9    | False   | False | Yes
P9    | False | True | Yes?*
P9    | True | False | Yes
P9    | True | True | No

* Should it be `Vulnerable`? 

CPU | L1D per Thread | RFI Flush | Vulnerable
--- | --- | --- | ---
P8 | False | False | Yes
P8 | False | True | Yes?**

** Should it be `Vulnerable`? IIRC P8 CPU's Threads share the L1D cache among the core, so they are always vulnerable.


```diff
diff --git a/arch/powerpc/kernel/security.c b/arch/powerpc/kernel/security.c
index b33bafb8fcea..d52852059ae3 100644
--- a/arch/powerpc/kernel/security.c
+++ b/arch/powerpc/kernel/security.c
@@ -130,26 +130,16 @@ ssize_t cpu_show_meltdown(struct device *dev, struct device_attribute *attr, cha

        thread_priv = security_ftr_enabled(SEC_FTR_L1D_THREAD_PRIV);

-       if (rfi_flush || thread_priv) {
-               struct seq_buf s;
-               seq_buf_init(&s, buf, PAGE_SIZE - 1);
-
-               seq_buf_printf(&s, "Mitigation: ");
-
-               if (rfi_flush)
-                       seq_buf_printf(&s, "RFI Flush");
-
-               if (rfi_flush && thread_priv)
-                       seq_buf_printf(&s, ", ");
-
+       if (rfi_flush) {
                if (thread_priv)
-                       seq_buf_printf(&s, "L1D private per thread");
-
-               seq_buf_printf(&s, "\n");
-
-               return s.len;
+                       return sprintf(buf, "Mitigation: RFI Flush, L1D private per thread\n");
+               else
+                       return sprintf(buf, "Vulnerable: RFI Flush\n");
        }

+       if (thread_priv)
+               return sprintf(buf, "Vulnerable: L1D private per thread\n");
+
        if (!security_ftr_enabled(SEC_FTR_L1D_FLUSH_HV) &&
            !security_ftr_enabled(SEC_FTR_L1D_FLUSH_PR))
                return sprintf(buf, "Not affected\n");
```

**Comment by mpe** [comment]
Yeah I guess it's arguable whether missing L1D private should report vulnerable. The way I think of it, the key attack is from user to kernel and the RFI flush mitigates that, so as long as the RFI flush is enabled we should print mitigation not vulnerable.

On the other hand if you have L1D private but **not** RFI flush, then you're vulnerable to attacks from user to kernel and so it should print vulnerable.

So I think what we want is:

CPU | L1D per Thread | RFI Flush | sysfs output
----| ----           | ---       | ---
P9  | False          | False     | Vulnerable
P9  | False          | True      | Mitigation: RFI Flush
P9  | True           | False     | Vulnerable, L1D private per thread
P9  | True           | True      | Mitigation: RFI Flush, L1D private per thread
P8  | False          | False     | Vulnerable
P8  | False          | True      | Mitigation: RFI Flush

**Comment by walbon** [comment]
I posted it in the mail list to be reviewed.

**Comment by mpe** [comment]
Fixed in torvalds/linux@4e706af3cd8e1d0503c25332b30cad33c97ed442 thanks @walbon :champagne:
